include/nuttx/lib/libvars.h

```diff
@@ -0,0 +1,76 @@
+/****************************************************************************
+ * include/nuttx/lib/libvars.h
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __INCLUDE_NUTTX_LIB_LIBVARS_H
+#define __INCLUDE_NUTTX_LIB_LIBVARS_H
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <stdbool.h>
+
+#include <nuttx/lib/getopt.h>
+
+/****************************************************************************
+ * Public Types
+ ****************************************************************************/
+
+#ifndef CONFIG_BUILD_KERNEL
+/* This structure encapsulates all task-specific variables needed by the C
+ * Library.  This structure is retained at the beginning of the main thread
+ * of and is accessed via a reference stored in the TLS of all threads in
+ * the task group.
+ *
+ * NOTE: task-specific variables are not needed in the KERNEL build.  Included
+ * that build mode, all global variables are inherently process-specific.
+ */
+
+struct libvars_s
+{
+  struct getopt_s lv_getopt; /* Globals used by getopt() */
+};
+#endif
```

Comment:
Would it be better if we held pointers here as well? This way we could only allocate each set of globals when they are actually used (so first call to getopt allocates it, otherwise not).

Reply:
The whole idea here is to allocate the memory directly on the main thread stack.  This structure represents that stack allocation.  What would the pointers do?  We would then have pointers to pointers to ... what?  Memory allocated somewhere outside of the stack?  That is exactly what we are trying to avoid here.

Reply:
In my mind we would allocate these globals on the heap on demand, holding pointers in this struct to that memory.
I was just concerned how much stack requirement for every task would increase even if they never use all libc functions that use those globals.

Reply:
I realized that the point of moving stuff to the stack is to avoid the deallocation problem. So I guess this is the cost of the solution.

Reply:
The is no application API to do that.  I don't think it is possible.  pthread_key_create() + pthread_setspecific() is the only POSIX compatible way to extend thread- (not task-) specific data.  I have no solution for the functionality you are interested in.  That, I think, would be a different PR and some long difficult discussion.

We have INVIOLABLES.md to help us with difficult, controversial decisions.  I am personally more concerned with the purity of the POSIX interface than with any application functionality.  It is the pure POSIX interface that distinguishes NuttX from other small OSs.

Reply:
If we toss standards and portability out the window, then I think we could implement something like pthread_key_create() and pthread_setspecific() and pthread_getspecific() to operate on an array of void * or uintptr_t in the task-specific data.  Only getopt() is addressed by this PR (as per the title and description).  Hopefully it provides a framework for other task-specific data, but that other task-specific data is out of scope of the current PR.

Reply:
I agree with current change, we can just add kconfig option to disable uneeded APIs which would avoid uneeded stack use.

Reply:
> I realized that the point of moving stuff to the stack is to avoid the deallocation problem. So I guess this is the cost of the solution.

Deallocation is not really a problem in the current implementation.  What I really wanted to avoid is small, persistent memory allocations that will fragment memory.

Reply:
> Maybe we can add options to compile out functions such as getopt() when not used, which would remove the relevant globals from the stack.

Possibly, but there are too many configuration options now.  getopt() requires only 24 bytes of storage.  Hardly worth the added complexity.

Reply:
> If we toss standards and portability out the window, then I think we could implement something like pthread_key_create() and pthread_setspecific() and pthread_getspecific() to operate on an array of void * or uintptr_t in the task-specific data. Only getopt() is addressed by this PR (as per the title and description). Hopefully it provides a framework for other task-specific data, but that other task-specific data is out of scope of the current PR.

Perhaps we could extend the prctl() commands for this kind of functionality?  prctl() is already non-standard and kind of makes sense to encapsulate task-wide controls of all kinds.  Currently it is used only for setting task names.

Reply:
> > Maybe we can add options to compile out functions such as getopt() when not used, which would remove the relevant globals from the stack.
> 
> Possibly, but there are too many configuration options now. getopt() requires only 24 bytes of storage. Hardly worth the added complexity.

Yes, each allocation has more cost than stack variables(min 16 bytes quantity, fragmentation and time for alloc/free). So it's better to embed small global variables like getopt used directly into stack.